src/Groups/GAPGroups.jl

```diff
@@ -0,0 +1,617 @@
+end
+
+function order(::Type{T}, x::Union{GAPGroupElem, Group}) where T<:Number
+   return T(order(x))
+end
+
+Base.exponent(x::Group) = GAP.Globals.Exponent(x.X)
+
+Base.length(x::Group) = order(x)
+
+function Base.rand(x::Group)
+   s = GAP.Globals.Random(x.X)
+   return group_element(x, s)
+end
+
+# one of the following should be non-parametric
+rand_pseudo(G::Group) = Base.rand(G)
+
+
+function _maxgroup(x::T, y::T) where T <: Group
+   if x == y
+     return x
+   elseif GAP.Globals.IsSubset(x.X, y.X)
+     return x
+   elseif GAP.Globals.IsSubset(y.X, x.X)
+     return y
+   else
+     error("Not yet implemented")
+   end
+end
+
+#We need a lattice of groups to implement this properly
+function _prod(x::T, y::T) where T <: GAPGroupElem
+  G = _maxgroup(parent(x), parent(y))
+  return group_element(G, x.X*y.X)
+end
+
+Base.:*(x::GAPGroupElem, y::GAPGroupElem) = _prod(x, y)
+
+function ==(x::Group, y::Group)
+   return x.X == y.X
+end
+
+function ==(x::T, y::T) where T <: GAPGroupElem
+   return x.X == y.X
+end
+
+"""
+    one(G::Group) -> x::GAPGroupElem{typeof(G)}
+Returns the one element of the group G.
+"""
+Base.one(x::Group) = group_element(x, GAP.Globals.Identity(x.X))
+
+"""
+    one(x::GAPGroupElement{T}) -> x::GAPGroupElem{T}
+Returns the one element of the parent of x.
+"""
+Base.one(x::GAPGroupElem) = one(parent(x))
+one!(x::GAPGroupElem) = one(parent(x))
+
+Base.show(io::IO, x::GAPGroupElem) =  print(io, GAP.gap_to_julia(GAP.Globals.StringView(x.X)))
+Base.show(io::IO, x::Group) = print(io, GAP.gap_to_julia(GAP.Globals.StringView(x.X)))
+
+Base.isone(x::GAPGroupElem) = x == one(parent(x))
+
+Base.inv(x::GAPGroupElem) = group_element(parent(x), GAP.Globals.Inverse(x.X))
+
+inv!(out::GAPGroupElem, x::GAPGroupElem) = inv(x)  #if needed later
+
+Base.:^(x::GAPGroupElem, y::Int) = group_element(parent(x), x.X ^ y)
+
+Base.:<(x::PermGroupElem, y::PermGroupElem) = x.X < y.X
+
+Base.:/(x::GAPGroupElem, y::GAPGroupElem) = x*y^-1
+
+mul(x::GAPGroupElem, y::GAPGroupElem) = x*y
+mul!(out::GAPGroupElem, x::GAPGroupElem, y::GAPGroupElem) = x*y
+
+div_right(x::GAPGroupElem, y::GAPGroupElem) = x*inv(y)
+div_left(x::GAPGroupElem, y::GAPGroupElem) = inv(y)*x
+div_right!(out::GAPGroupElem, x::GAPGroupElem, y::GAPGroupElem) = x*inv(y)
+div_left!(out::GAPGroupElem, x::GAPGroupElem, y::GAPGroupElem) = inv(y)*x
+
+Base.conj(x::GAPGroupElem, y::GAPGroupElem) = x^y
+Base.conj!(out::GAPGroupElem, x::GAPGroupElem, y::GAPGroupElem) = x^y
+
+comm(x::GAPGroupElem, y::GAPGroupElem) = x^-1*x^y
+comm!(out::GAPGroupElem, x::GAPGroupElem, y::GAPGroupElem) = x^-1*x^y
+
+function Base.iterate(G::Group)
+  L=GAP.Globals.Iterator(G.X)
+  if GAP.Globals.IsDoneIterator(L)
+    return nothing
+  end
+  i = GAP.Globals.NextIterator(L)
+  return group_element(G, i), L
+end
+
+function Base.iterate(G::Group, state)
+  if GAP.Globals.IsDoneIterator(state)
+    return nothing
+  end
+  i = GAP.Globals.NextIterator(state)
+  return group_element(G, i), state
+end
```

Comment:
Hmm, I guess we could also provide `Base.IteratorSize` methods for `PermGroup` and `PcGroup` at least (returning `HasLength()` for both), and also `Base.IteratorEltype`.

Reply:
`HasLength()` is the default, but it's always nice to be explicit.

Reply:
Yeah I'd actually change the default for our groups to "unknown", to be on the safe side (i.e. be possibly non-optimal but always correct)